metron-platform/metron-common/src/main/java/org/apache/metron/common/configuration/ConfigurationsUtils.java

```diff
@@ -343,25 +343,57 @@ public static void uploadConfigsToZookeeper(String rootFilePath, CuratorFramewor
    * @param type config type to upload configs for
    * @param configName specific config under the specified config type
    */
-  public static void uploadConfigsToZookeeper(String rootFilePath, CuratorFramework client,
-      ConfigurationType type, Optional<String> configName) throws Exception {
+  public static void uploadConfigsToZookeeper(
+          String rootFilePath,
+          CuratorFramework client,
+          ConfigurationType type,
+          Optional<String> configName) throws Exception {
+
     switch (type) {
+
       case GLOBAL:
         final byte[] globalConfig = readGlobalConfigFromFile(rootFilePath);
         if (globalConfig.length > 0) {
           setupStellarStatically(client, Optional.of(new String(globalConfig)));
           writeGlobalConfigToZookeeper(globalConfig, client);
         }
         break;
-      case PARSER: // intentional pass-through
-      case ENRICHMENT: // intentional pass-through
-      case INDEXING:
-        Map<String, byte[]> sensorIndexingConfigs = readSensorConfigsFromFile(rootFilePath, type,
-            configName);
-        for (String sensorType : sensorIndexingConfigs.keySet()) {
-          writeConfigToZookeeper(type, configName, sensorIndexingConfigs.get(sensorType), client);
+
+      case PARSER: {
+        Map<String, byte[]> configs = readSensorConfigsFromFile(rootFilePath, PARSER, configName);
```

Comment:
These cases look like they are cut and pasted which seems like code smell to me and might be a maintenance issue.  Can we extract the common code for Parser, Enrichment, and Indexing into a separate function that is called here?  Perhaps something like:
```
void writeSensorConfigs(ConfigurationType type, Optional<String> configName, BiFunction<String, byte[], Void> callback) {
   Map<String, byte[]> configs = readSensorConfigsFromFile(rootFilePath, type, configName);
  for(String sensorType : configs.keySet()) {
    byte[] configData = configs.get(sensorType);
    callback.apply(sensorType, configData);
  }
}
```
which could be called from this case via `writeSensorConfigs(PARSER, configName, (sensorType, configData) -> writeSensorParserConfigToZookeeper(sensorType, configData, client));`

Take the above as a very rough suggestion and you can feel to abstract it however you wish.

Reply:
The other option here is to add a new method called `writeSensorConfigToZookeeper(ConfigurationType type, String sensorType, byte[] configData, CuratorFramework client)` that calls the appropriate `writeSensorXConfigToZookeeper` call.

Reply:
Oh, an even better option (IMO) would be to add a `writeSensorConfigToZookeeper(String sensorType, byte[] configData, CuratorFramework client)` method to `ConfigurationType`.  Anyway, I'll stop ;)

Reply:
Yes, I had the same thought and tried for a bit to refactor it.  I landed on this because the various other ways to do this either (1) seemed more complex and less obvious as to what we are actually doing here or (2) lead down a path of heavy refactoring of ConfigurationUtils.  Both of which i wanted to avoid

Reply:
Your first suggestion with the callback just seems less obvious and clear to me (IMHO).

I'll try and think through your 2nd and 3rd suggestions, but (at least right now) I'm not seeing something that doesn't add more complexity to ConfigurationUtils.  

But clearly I could be totally wrong here.  Let me think on it.

Reply:
Sounds good.  Just so I'm clear about my preferred approach (option 3 there), I went ahead and implemented it in a [PR](https://github.com/nickwallen/metron/pull/12) against your PR so you can see what I was talking about.

Reply:
I remember struggling through this class when applying the patching mechanism, so I want to start by saying I'm a big fan of this work and conversation around how to best improve and simplify this class. @cestella I agree with @nickwallen about option 1. Compared to your PR on this PR (sounds like an XZibit meme...) for option 3, I think it's much less intuitive to read. I took a look through it and that would be my preference for the way to go. It's classic strategy pattern through composition, and it eliminates the duplication we see below. I tend to agree with the comments on that PR about eventually pulling those implementations into their own classes. That would make the code much more readable. And I think it probably puts us in a better place for cleaning up the config utils class or removing/renaming it altogether. The only part of this that is still a little bit of a smell to me is that the global config and profiler throw unsupported operation exceptions. I think this is a big enough improvement that I'd be willing to let that slide for now, but ideally the global config would exist similarly to the other config. e.g. instead of this:

- config/zookeeper/
  - global.json
  - profiler.json
  - parsers/
    - ...
  - enrichment/
    - ...
  - indexing/
    - ...
  - .... etc.

You'd give the global config and profiler a directory as well, which would enable a consistent layout on the local FS and in Zookeeper, and a consistent interface for these commands. 

- config/zookeeper/
  - global/
    - global.json
  - profiler/
    - profiler.json
  - parsers/
    - ...

Then this
```
public interface ConfigurationOperations {
...
void writeSensorConfigToZookeeper(String sensorType, byte[] configData, CuratorFramework client) throws Exception;
...
```

could become this

```
public interface ConfigurationOperations {
...
void writeConfigToZookeeper(String type, byte[] configData, CuratorFramework client) throws Exception;
```

I'm not sure it's desirable or not, but you could then theoretically provide multiple configs in global/ and profiler/, which is consistent with the parser/, enrichment/, and indexing/ configs.

Reply:
One more thought - the code for option 3 does add more code than the copy/paste, but in spite of that I think it puts us closer to being able to delete a lot of the redundant and/or dead code in config utils. So I think it's a net positive.

Reply:
@cestella I submitted a PR against your PR on this PR - https://github.com/cestella/incubator-metron/pull/8

Reply:
@mmiklavc I'm ok with that modification if we decide to go this route.

Reply:
I like the refactoring, guys.  It's looking food.  Can't believe we have a PR of a PR of this PR.  One more level and I think we hit Inception's limbo.

I'd be open to including those changes here or even just creating a separate PR for the refactoring work as a follow-on to this.  That refactoring adds a good bit of code in itself.  Maybe we could even tackle some further ConfigurationUtils clean-up in that same PR.

But If we want @cestella and @mmiklavc 's refactoring to go in with this PR, then we need @cestella to do the first "kick" and merge @mmiklavc's PR.

Reply:
Haha, I know. I couldn't resist seeing if GitHub actually allowed this sort of thing, and well, it was actually pretty straightforward. I don't care if we pull it all together or submit them separately. It sounds like we're all in agreement that the refactoring is a good idea, so if it was split them Casey's interface refactoring would go in immediately following with the necessary votes. I'll leave it to you both to decide.

Reply:
The meta-PR is merged